overlord/state/taskrunner.go

```diff
@@ -0,0 +1,156 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package state
+
+import (
+	"sync"
+
+	"gopkg.in/tomb.v2"
+)
+
+// HandlerFunc is the type of function for the hanlders
+type HandlerFunc func(task *Task) error
+
+// TaskRunner controls the running of goroutines to execute known task kinds.
+type TaskRunner struct {
+	state *State
+
+	// locking
+	mu       sync.Mutex
+	handlers map[string]HandlerFunc
+
+	// go-routines lifecycle
+	tombs map[string]*tomb.Tomb
+}
+
+// NewTaskRunner creates a new TaskRunner
+func NewTaskRunner(s *State) *TaskRunner {
+	return &TaskRunner{
+		state:    s,
+		handlers: make(map[string]HandlerFunc),
+		tombs:    make(map[string]*tomb.Tomb),
+	}
+}
+
+// AddHandler registers the function to concurrently call for handling
+// tasks of the given kind.
+func (r *TaskRunner) AddHandler(kind string, fn HandlerFunc) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.handlers[kind] = fn
+}
+
+// Handlers returns the map of name/handler functions
+func (r *TaskRunner) Handlers() map[string]HandlerFunc {
+	return r.handlers
+}
+
+// run must be called with the state lock in place
```

Comment:
run can probably take the Task and not fish it out again from state? and just do taskID = task.ID() inside